### PR TITLE
Fix collision flags and path finding

### PR DESCRIPTION
**Pull request by garyttierney**
- Adds support for wall objects and other types to GameObjectDecoder.
- Adds support for the diagonal collision flags from the landscape and
  object maps, emulating what is done in the 317 client release.
- Adds support for raycasting through the world for collisions.
- Verifies that steps in the walking queue are traversable from the
  current position.
- Changes PathfindingAlgorithm implementations to use World#intersects
  instead of Region#traversable.

Still a WIP. TODO: 
- [x] remove unrelated code from branch (oops)
- [x] <s>add support for uneven (?) collision paths to intersectable (e.g, dx=15,dy=17)</s> - the simple path finder should get diagonal support and be used instead
- [x] possibly clean up how tiles are flaged

### CODE REVIEW

**Comment by kylestev** [comment]
@Major- wouldn't it make sense to setup some CI for style checking PRs?

**Comment by kylestev** [comment]
https://github.com/checkstyle/checkstyle for a code style checker with Travis support. I would be happy to write the style settings to match your current standards and then all your need to do is sign this repo up for Travis.

**Comment by Major-** [comment]
@kylestev yeah it probably would, I experimented with PMD a while back but it was very annoying and generally useless as I ended up disabling loads of checks because of false positives. Do you know if checkstyle is any better?

**Comment by kylestev** [comment]
Haven't used it recently. It wasn't too much of a pain though. I'll take a
look at it when I get home in 30 minutes and get back to you.

**Comment by Major-** [comment]
Mostly nitpicking, LGTM

Needs rebase

**Comment by garyttierney** [comment]
ping @Major-, what's blocking this?

**Comment by Major-** [comment]
Nothing (apart from that it needs a rebase)

**Comment by WizardJesse1** [comment]
Currently found an issue with this

When you are in the scimitar shop u can not leave due to curtains. the curtains are open so i am not sure why u cant access it 

coords to scimmy shop = 3288 3189

**Comment by Major-** [comment]
cc @garyttierney what's going on with this? any idea about jesse's comment?

**Comment by WizardJesse1** [comment]
@Major- i already fixed i just wanted to show gary first

**Comment by garyttierney** [comment]
@Major- issue with some of the `GameObjectDecoder`. Going to pull a fix from @WizardJesse1 and work on this soon.

**Comment by nstroupr** [comment]
At the moment the pathing gets stuck when trying to go through a doorway.

**Comment by garyttierney** [comment]
@nstroupr that is an issue with this current tree, since we don't update the collision matrices when an object is added or removed to a region. My local tree accounts for this so I'll work on it a bit when I merge following.

File: game/src/main/org/apollo/game/model/Direction.java [hunk 1]
```diff
@@ -58,6 +58,17 @@
 	public static final Direction[] EMPTY_DIRECTION_ARRAY = new Direction[0];
 
 	/**
+	 * An array of directions without any diagonal directions.
+	 */
+	public static Direction[] NESW = { NORTH, EAST, SOUTH, WEST };
```
Comment (ryleykimmel) [inline game/src/main/org/apollo/game/model/Direction.java]:
```java
final
```
here and below?

**Comment by Lmctruck30** [comment]
@garyttierney @ryleykimmel When you changed from Region#traversable to World#intersects now doors do not work. Needs to removed the collision flag from the matrix ie change flag to 0 or have a temporary opening for doors. As it stands now you can not walk past an open door without creating an edit. What I have been testing is a handler that controls all the doors in the world and will block and open traversal paths.

**Comment by ryleykimmel** [comment]
@Lmctruck30 read the comments above, they discussed the issue and @garyttierney has patched it in his local tree.

**Comment by garyttierney** [comment]
See https://github.com/apollo-rsps/apollo/pull/302#issuecomment-263251090

**Comment by garyttierney** [comment]
@WizardJesse1 ping, fancy checking if you can reproduce any of your issues with this tree?

**Pull request merged**